### PR TITLE
Optimized function schema message for `ToolMetadata`

### PR DESCRIPTION
**Pull request by azurewtl**
# Description
Let's say we define a super simple schema
```python
class MySchema(BaseModel):
    input: Optional[str]
```
## Before
`ToolMetadata.fn_schema_str` returns:
```python
{'title': 'FnSchema', 'description': 'Default tool function Schema.', 'type': 'object', 'properties': {'input': {'title': 'Input', 'type': 'string'}}}
```
***Problem:*** It contain too many information such as `title`, and it's not a valid `Json`, because it simply produced by `str(model.schema())`
## After
`ToolMetadata.fn_schema_str` returns:
```json
{"type": "object", "properties": {"input": {"title": "Input", "type": "string"}}}
```

## Usage Pattern
use it in query_engine
```python
query_engine_tools = [
    RetrieverTool(
        retriever=vs_retriever,
        metadata=ToolMetadata(
            name="useless tool", 
            description="super useless",
             fn_schema=MySchema
        ),
    ),
]
```
when we prompt tool to LLM by
```python
def get_react_tool_descriptions(tools: Sequence[BaseTool]) -> List[str]:
    """Tool."""
    tool_descs = []
    for tool in tools:
        tool_desc = (
            f"> Tool Name: {tool.metadata.name}\n"
            f"Tool Description: {tool.metadata.description}\n"
            f"Tool Args: {tool.metadata.fn_schema_str}\n"
        )
        tool_descs.append(tool_desc)
    return tool_descs
```
`tool.metadata.fn_schema_str` is the description of schema that LLM use to understand how to call the tool

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by logan-markewich** [comment]
Does this break compatibility with OpenAI tool/function calling though?

File: llama_index/tools/types.py [hunk 1]
```diff
@@ -20,12 +21,31 @@ class ToolMetadata:
...
+                    "input": {"title": "input query string", "type": "string"},
+                },
+                "required": ["input"],
+            }
+        else:
+            parameters = self.fn_schema.schema()
+            parameters = {
+                k: v
+                for k, v in parameters.items()
+                if k in ["type", "properties", "required"]
```
Comment (azurewtl) [inline llama_index/tools/types.py:38]: only keep three keys that openai function require

**Comment by azurewtl** [comment]
> Does this break compatibility with OpenAI tool/function calling though?

Reference:[openai documentation](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_call_functions_with_chat_models.ipynb)
Yes, the three key I kept is based on above document in `function`.`parameters`

```python
tools = [
  {
      "type": "function",
      "function": {
          "name": "get_current_weather",
          "description": "Get the current weather",
          "parameters": {
              "type": "object",
              "properties": {
                  "location": {
                      "type": "string",
                      "description": "The city and state, e.g. San Francisco, CA",
                  },
                  "format": {
                      "type": "string",
                      "enum": ["celsius", "fahrenheit"],
                      "description": "The temperature unit to use. Infer this from the users location.",
                  },
              },
              "required": ["location", "format"],
          },
      }
  }
```

File: llama_index/tools/types.py [hunk 2]
```diff
@@ -20,12 +21,31 @@ class ToolMetadata:
...
+            }
+        return parameters
+
     @property
     def fn_schema_str(self) -> str:
         """Get fn schema as string."""
         if self.fn_schema is None:
             raise ValueError("fn_schema is None.")
-        return str(self.fn_schema.schema())
```
Comment (azurewtl) [inline llama_index/tools/types.py:28]: line 28: str(self.fn_schema.schema()) is not a valid json

**Comment by azurewtl** [comment]
@logan-markewich can I get merged?
otherwise phoenix callback gets error, this PR fits it
```
**************************************************
Traceback (most recent call last):
  File "/Users/azure/Documents/Workspace/Datasci/lib/python3.10/site-packages/phoenix/utilities/error_handling.py", line 34, in wrapper
    return func(*args, **kwargs)
  File "/Users/azure/Documents/Workspace/Datasci/lib/python3.10/site-packages/phoenix/trace/llama_index/callback.py", line 266, in on_event_start
    payload_to_semantic_attributes(event_type, payload),
  File "/Users/azure/Documents/Workspace/Datasci/lib/python3.10/site-packages/phoenix/trace/llama_index/callback.py", line 194, in payload_to_semantic_attributes
    attributes[TOOL_PARAMETERS] = tool_metadata.to_openai_function()["parameters"]
KeyError: 'parameters'
```

**Comment by sandangel** [comment]
Hi, Is there any update on this PR? I think this looks good. THe current implementation is not producing the json format.

**Comment by logan-markewich** [comment]
@azurewtl wow sorry, I let this PR slip. Thanks for the reminder @sandangel 

Just tested this out a bit with a few various tools (LLMTextCompletionProgram, ReActAgent), seemed to work well with both gpt-3.5-turbo and gpt-3.5-turbo-instruct

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**